designs/node-overlay-dra.md

```diff
@@ -0,0 +1,564 @@
+| `spec.devices[].taints`                   | Alpha (Device Taints)                                                                                                                                                                    |
+
+New fields would be evaluated on a case-by-case basis as they are added to the upstream schema. Additionally, new fields
+may be introduced to support functionality specific to Karpenter. An example of this can be found in the section on
+[supporting relative topology requirements](#supporting-relative-topology-requirements). A complete description of the
+proposed `ResourceSliceTemplate` schema can be found in the appendix.
+
+### Supporting Relative Topology Requirements
+
+One of the main additional capabilities that DRA enables is non-node local device consumption. Drivers can define the
+set of nodes a device can be consumed from using `spec.nodeSelector` on the `ResourceSlice`. Generally speaking, this
+topology requirement is going to be **relative** to the device's "host node." Going forward, we'll consider an example
+of a device which is only available within the same zone as the host node.
+
+When a DRA driver creates `ResourceSlices` for such a device, it already knows the zone the node has been provisioned in
+and can inject that zonal requirement into the `ResourceSlice's` selector. On the other hand, the user does not
+necessarily know the zone that Karpenter will provision the node into. They could create a `NodeOverlay` per topology
+domain, but this becomes untennable with a large number of domains. This problem statement can be generalized: how does
+a user express a topology constraint when the domain is unresolved?
+
+There is prior art in Kubernetes for expressing topological constraints based on unresolved domains:
+[`matchLabelKeys` for pod affinity and topology spread constraints](https://kubernetes.io/blog/2024/08/16/matchlabelkeys-podaffinity/).
+The motivating example for this feature was very similar: users need a way to express constraints based on a pod's
+template hash, but that hash won't be resolved until the pod is created. Therefore, the user can't express this directly
+in the pod template. This is the same issue we face with DRA: we know the topology we want to express the constraint on
+(zone) but won't know the domain until the node is provisioned. This RFC proposes adding `matchLabelKeys` to
+`spec.nodeSelector` on the `ResourceClaimTemplate` where it will serve the same purpose as it does in the pod spec.
+
+An example of configuring a zonal selector using `matchLabelKeys`:
+
+```yaml
+kind: NodeOverlay
+apiVersion: karpenter.sh/v1alpha1
+spec:
+  resourceSliceTemplates:
+    - spec:
+        nodeSelector:
+          nodeSelectorTerms:
+            - matchLabelKeys:
+                - "topology.kubernetes.io/zone"
+```
+
+**Note:** This RFC does **not** propose extending the upstream `ResourceSlice` schema with `matchLabelKeys`. Although we
```

Comment:
So the user now has to plan for having ResourceSlice on the cluster for kube-scheduler and configuring overlays to make it work with Karpenter? Cant we directly use the CR the driver creates and read specific fields instead of maintaining one here?

Reply:
The reason we're adding `ResourceSlices` to `NodeOverlays` is because they **haven't** been created on the cluster yet. We need to be able to anticipate what `ResourceSlices` will be generated for an instance type before we launch it to understand if that instance type will be able to support the pod's requests.

Reply:
For nodes that have already been created, Karpenter will use the existing `ResourceSlices`.